tests/5.1/target/test_target_memcpy_async_no_obj.c

```diff
@@ -0,0 +1,69 @@
+//===--- test_target_memcpy_async_no_obj.c ----------------------------===//
+//
+//  OpenMP API Version 5.1 Aug 2021
+//
+//  Inspired from OpenMP 5.1 Examples Doc, 5.16.4 & 8.9
+//  This test utilizes the omp_target_memcpy_async construct to
+//  allocate memory on the device asynchronously. The construct
+//  uses 0 for 'depobj_count', so that the clause is not dependent
+//  and memory is therefore copied synchronously.
+//
+////===----------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+#include <math.h>
+
+#define N 1024
+
+int errors, i;
+
+int test_target_memcpy_async_no_obj() {
+
+    int h, t, i;
+    errors = 0;
+    double *mem;
+    double *mem_dev_cpy;
+    h = omp_get_initial_device();
+    t = omp_get_default_device();
+
+    OMPVV_TEST_AND_SET_VERBOSE(errors, omp_get_num_devices() < 1 || t < 0);
+
+    //allocate space on host & target
+    mem = (double *)malloc( sizeof(double)*N);
+    mem_dev_cpy = (double *)omp_target_alloc( sizeof(double)*N, t);
+
+    OMPVV_TEST_AND_SET_VERBOSE(errors, mem_dev_cpy == NULL);
+
+    /* copy to target */
+    omp_target_memcpy_async(mem_dev_cpy, mem, sizeof(double)*N,
+                                0,          0,
+                                t,          h,
+                                0,          0);
+
+    #pragma omp target is_device_ptr(mem_dev_cpy) device(t)
+    #pragma omp teams distribute parallel for
+    for(i = 0;i<N;i++){
+        mem_dev_cpy[i] = i*2; // initialize data on device
+    }
+    /* copy to host */
+    omp_target_memcpy_async(mem, mem_dev_cpy, sizeof(double)*N,
+                                0,          0,
+                                h,          t,
+                                0,          0);
```

Comment:
```suggestion
                                0,          NULL);
```
makes reading the code easier.

However, my main problem is that this code assumes: _No depend object → code runs synchronously._

But my reading of the spec is that it just runs asynchronously and the only synchronization barrier is then the end of the current task.

OpenMP 5.2 spec has:
> **Binding** – The binding task set for an omp_target_memcpy_async region is the generating task, which is the _target task_ generated by the call to the omp_target_memcpy_async routine.
>
> **Effect** – [...] The **omp_target_memcpy_async** routine executes as if part of a target task that is generated by the call to the routine and for which execution may be deferred. Task dependences are expressed with zero or more OpenMP depend objects. The dependences are specified by passing the number of depend objects followed by an array of the objects. The generated target task is not a dependent task if the program passes in a count of zero for _depobj_count._ _depobj_list_ is ignored if the value of _depobj_count_ is zero.

Reading/thinking about it, I think something like the following would work; to be inserted after both `omp_target_memcpy_async`
* `omp taskwait`  (I think this one is better)
* `omp barrier`

Reply:
I am still struggling to understand tasks in OMP.... what would the 'current task' be in this case.. is it explicit? @tob2 

Anyways, I will add `omp taskwait` as it can't hurt

Reply:
> what would the 'current task' be in this case

I think the following applies (emphasis is mine):

"An OpenMP program begins as a single thread of execution, called an initial thread. An initial thread executes sequentially, as if the code encountered is part of an **implicit task region,** called an **initial task region,** that is **generated by** the **implicit parallel region surrounding the whole program.**" (From OpenMP 5.2, 1.3 Execution Model)